File: solutions/exercise-03/src/main/java/io/miragon/training/application/service/SendConfirmationMailService.java
package io.miragon.training.application.service;

import io.miragon.training.application.port.inbound.SendConfirmationMailUseCase;
import io.miragon.training.application.port.outbound.MembershipRepository;
import io.miragon.training.domain.MembershipId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class SendConfirmationMailService implements SendConfirmationMailUseCase {

    private static final Logger log = LoggerFactory.getLogger(SendConfirmationMailService.class);

    private final MembershipRepository repository;

    public SendConfirmationMailService(MembershipRepository repository) {
        this.repository = repository;
    }

    @Override
    public void sendConfirmationMail(MembershipId membershipId) {
        var membership = repository.find(membershipId);
        log.info("Sending confirmation mail to {} (membershipId={})", membership.email().value(), membershipId.value());
    }
}

File: solutions/exercise-2/src/main/kotlin/io/miragon/training/application/service/SendConfirmationMailService.kt
package io.miragon.training.application.service

import io.github.oshai.kotlinlogging.KotlinLogging
import io.miragon.training.application.port.inbound.SendConfirmationMailUseCase
import io.miragon.training.application.port.outbound.MembershipRepository
import io.miragon.training.domain.MembershipId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class SendConfirmationMailService(
    private val repository: MembershipRepository,
) : SendConfirmationMailUseCase {

    private val log = KotlinLogging.logger {}

    override fun sendConfirmationMail(membershipId: MembershipId) {
        val membership = repository.find(membershipId)
        log.info { "Sending confirmation mail to ${membership.email.value} (membershipId=${membershipId.value})" }
    }
}
